# Homework

In [1]:
#importing libraries
import pandas as pd
import numpy as np

import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.feature_extraction import DictVectorizer
from sklearn.linear_model import LinearRegression, Lasso, Ridge
from sklearn.metrics import mean_squared_error

## 1. Read the data for January. How many columns are there?

In [2]:
# reading the january dataset.
jan_df = pd.read_parquet("https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-01.parquet")
jan_df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1         2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2         2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3         1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4         2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           0.97         1.0                  N           161           141   
1           1.10         1.0                  N            43           237   
2           2.51         1.0                  N            48           238   
3           1.90         1.0                  N           138             7   
4           1.43         1.0                  N           107            79   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          9.3   1.00      0.5        0.00           0.0   
1             1          7.9   1.00      0.5        4.00           0.0   
2             1         14.9   1.00      0.5       15.00           0.0   
3             1         12.1   7.25      0.5        0.00           0.0   
4             1         11.4   1.00      0.5        3.28           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  
0                    1.0         14.30                   2.5         0.00  
1                    1.0         16.90                   2.5         0.00  
2                    1.0         34.90                   2.5         0.00  
3                    1.0         20.85                   0.0         1.25  
4                    1.0         19.68                   2.5         0.00

In [3]:
# nums of rows and columns
jan_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3066766 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[ns]
 2   tpep_dropoff_datetime  datetime64[ns]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee           

## Answer 1

* There are 19 columns in the january dataset.

## 2. What's the standard deviation of the trips duration in January?

In [4]:
# computing the duraion variable
jan_df["duration"] = jan_df.tpep_dropoff_datetime - jan_df.tpep_pickup_datetime

 # convert duration column to mins
jan_df.duration = jan_df["duration"].apply(lambda td: td.total_seconds()/ 60)

In [5]:
# estiamting the standard deviation of duraion variable
jan_df.describe().iloc[:, -1]

count    3.066766e+06
mean     1.566900e+01
std      4.259435e+01
min     -2.920000e+01
25%      7.116667e+00
50%      1.151667e+01
75%      1.830000e+01
max      1.002918e+04
Name: duration, dtype: float64

## Answer 2

* The standard deviation for duration is 42.59

## 3. What fraction of the records left after you dropped the outliers?

In [6]:
# fraction of recorda left dropping outliers
((jan_df.duration >= 1) & (jan_df.duration <= 60)).mean() * 100

98.1220282212598

In [7]:
# filtering the dataset to remove outliers
jan_df = jan_df[(jan_df.duration >= 1) & (jan_df.duration <= 60)]

jan_df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1         2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2         2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3         1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4         2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           0.97         1.0                  N           161           141   
1           1.10         1.0                  N            43           237   
2           2.51         1.0                  N            48           238   
3           1.90         1.0                  N           138             7   
4           1.43         1.0                  N           107            79   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          9.3   1.00      0.5        0.00           0.0   
1             1          7.9   1.00      0.5        4.00           0.0   
2             1         14.9   1.00      0.5       15.00           0.0   
3             1         12.1   7.25      0.5        0.00           0.0   
4             1         11.4   1.00      0.5        3.28           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  \
0                    1.0         14.30                   2.5         0.00   
1                    1.0         16.90                   2.5         0.00   
2                    1.0         34.90                   2.5         0.00   
3                    1.0         20.85                   0.0         1.25   
4                    1.0         19.68                   2.5         0.00   

    duration  
0   8.433333  
1   6.316667  
2  12.750000  
3   9.616667  
4  10.833333

## Answer 3

* The fraction left dropping outliers is 98.12%

## 4.What's the dimensionality of this matrix (number of columns)?

In [8]:
# selecting the features
jan_cat = ["PULocationID", "DOLocationID"]

#convertng the categorical features to str datatype
jan_df[jan_cat] = jan_df[jan_cat].astype(str)

# intializing DictVectorzer
dv = DictVectorizer()

# converting the catorical train data to dictionary data type
jan_train_dict = jan_df[jan_cat].to_dict(orient = "records")
jan_X_train = dv.fit_transform(jan_train_dict)

In [9]:
# number of columns in the jan data after fitting with dv
len(dv.feature_names_)

515

## Answer 4

* The dimensionality of this matrix (number of columns) is 515

## 5. What's the RMSE on train?

In [10]:
# target variable for january data
jan_target = "duration"
jan_y_train = jan_df[jan_target].values

In [11]:
# intializing linear regression moodel
lr = LinearRegression()
# fitting the train data into the linear regression model
lr.fit(jan_X_train, jan_y_train)

# predicting the validation target
jan_y_pred = lr.predict(jan_X_train)

# estimating the error between the val original target variable with the val predicted  target variable using rmse.
mean_squared_error(jan_y_train, jan_y_pred, squared = False)

7.649261027855596

## Answer 5

* The RMSE on train is 7.65

## 6. What's the RMSE on validation?

In [12]:
# Designing a function that reads the dataframe and perform preprocessing of the data
def read_dataframe(filename):

    # read the dataframe
    df = pd.read_parquet(filename)

    # create a target variable called duration
    df["duration"] = df.tpep_dropoff_datetime - df.tpep_pickup_datetime
    # convert duration column to mins
    df.duration = df["duration"].apply(lambda td: td.total_seconds()/ 60)

    # filtering the dataset to remove outliers
    df = df[(df.duration >= 1) & (df.duration <= 60)]

    # selecting the features
    categorical = ["PULocationID", "DOLocationID"]
    df[categorical] = df[categorical].astype(str)

    return df

In [13]:
feb_val_df = read_dataframe("https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-02.parquet")

In [14]:
feb_val_df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2023-02-01 00:32:53   2023-02-01 00:34:34              2.0   
3         1  2023-02-01 00:29:33   2023-02-01 01:01:38              0.0   
4         2  2023-02-01 00:12:28   2023-02-01 00:25:46              1.0   
5         1  2023-02-01 00:52:40   2023-02-01 01:07:18              1.0   
6         1  2023-02-01 00:12:39   2023-02-01 00:40:36              1.0   

   trip_distance  RatecodeID store_and_fwd_flag PULocationID DOLocationID  \
0           0.30         1.0                  N          142          163   
3          18.80         1.0                  N          132           26   
4           3.22         1.0                  N          161          145   
5           5.10         1.0                  N          148          236   
6           8.90         1.0                  N          137          244   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          4.4   3.50      0.5        0.00           0.0   
3             1         70.9   2.25      0.5        0.00           0.0   
4             1         17.0   1.00      0.5        3.30           0.0   
5             1         21.9   3.50      0.5        5.35           0.0   
6             1         41.5   3.50      0.5        3.50           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  Airport_fee  \
0                    1.0          9.40                   2.5         0.00   
3                    1.0         74.65                   0.0         1.25   
4                    1.0         25.30                   2.5         0.00   
5                    1.0         32.25                   2.5         0.00   
6                    1.0         50.00                   2.5         0.00   

    duration  
0   1.683333  
3  32.083333  
4  13.300000  
5  14.633333  
6  27.950000

In [15]:
# selecting the features
feb_cat = ["PULocationID", "DOLocationID"]

# converting the catorical val data to dictionary data type
feb_val_dict = feb_val_df[feb_cat].to_dict(orient = "records")
feb_X_val = dv.transform(feb_val_dict)

In [16]:
# target variable
feb_target = "duration"

feb_y_val = feb_val_df[feb_target].values

In [17]:
# predicting the validation target
feb_y_pred = lr.predict(feb_X_val)

# estimating the error between the val original target variable with the val predicted  target variable using rmse.
mean_squared_error(feb_y_val, feb_y_pred, squared = False)

7.8118325109416915

## Answer 5

* The RMSE on validation data is 7.81